File: src/controllers/user_controller.go
package controllers

import (
	"strconv"

	"github.com/gofiber/fiber/v2"
	"gorm.io/gorm"

	"api21/src/models"
)

// UserController handles user-related HTTP requests
type UserController struct{}

// NewUserController creates a new UserController instance
func NewUserController() *UserController {
	return &UserController{}
}

// GetUsers handles GET /api/users - returns all users
func (uc *UserController) GetUsers(c *fiber.Ctx) error {
	users, err := models.GetAllUsers()
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to retrieve users",
			"error":   err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"message": "Users retrieved successfully",
		"data":    users,
		"count":   len(users),
	})
}

// GetUser handles GET /api/users/:id - returns a single user by ID
func (uc *UserController) GetUser(c *fiber.Ctx) error {
	idParam := c.Params("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid user ID",
			"error":   err.Error(),
		})
	}

	user, err := models.GetUserByID(uint(id))
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
				"success": false,
				"message": "User not found",
			})
		}
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to retrieve user",
			"error":   err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"message": "User retrieved successfully",
		"data":    user,
	})
}

// CreateUser handles POST /api/users - creates a new user
func (uc *UserController) CreateUser(c *fiber.Ctx) error {
	var requestData struct {
		Name  string `json:"name"`
		Email string `json:"email"`
	}

	if err := c.BodyParser(&requestData); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
			"error":   err.Error(),
		})
	}

	if requestData.Name == "" || requestData.Email == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Name and email are required",
		})
	}

	user := models.NewUser(requestData.Name, requestData.Email)
	if err := user.CreateUser(); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to create user",
			"error":   err.Error(),
		})
	}

	return c.Status(fiber.StatusCreated).JSON(fiber.Map{
		"success": true,
		"message": "User created successfully",
		"data":    user,
	})
}

// UpdateUser handles PUT /api/users/:id - updates an existing user
func (uc *UserController) UpdateUser(c *fiber.Ctx) error {
	idParam := c.Params("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid user ID",
			"error":   err.Error(),
		})
	}

	user, err := models.GetUserByID(uint(id))
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
				"success": false,
				"message": "User not found",
			})
		}
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to retrieve user",
			"error":   err.Error(),
		})
	}

	var requestData struct {
		Name  string `json:"name"`
		Email string `json:"email"`
	}

	if err := c.BodyParser(&requestData); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
			"error":   err.Error(),
		})
	}

	if requestData.Name != "" {
		user.Name = requestData.Name
	}
	if requestData.Email != "" {
		user.Email = requestData.Email
	}

	if err := user.UpdateUser(); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to update user",
			"error":   err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"message": "User updated successfully",
		"data":    user,
	})
}

// DeleteUser handles DELETE /api/users/:id - deletes a user
func (uc *UserController) DeleteUser(c *fiber.Ctx) error {
	idParam := c.Params("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid user ID",
			"error":   err.Error(),
		})
	}

	user, err := models.GetUserByID(uint(id))
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
				"success": false,
				"message": "User not found",
			})
		}
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to retrieve user",
			"error":   err.Error(),
		})
	}

	if err := user.DeleteUser(); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Failed to delete user",
			"error":   err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"message": "User deleted successfully",
	})
}

File: src/controllers/userController.ts
import { Request, Response } from 'express';
import { getDbPool } from '../infrastructure/database.js';

export const getUsers = async (req: Request, res: Response) => {
  try {
    const db = getDbPool();
    const result = await db.query('SELECT id, name, email, created_at, updated_at FROM users ORDER BY id ASC');
    res.json({
      status: 'success',
      data: result.rows,
    });
  } catch (err: any) {
    console.error('[UserController] getUsers error:', err);
    res.status(500).json({ status: 'error', message: 'Failed to fetch users' });
  }
};

export const getUserById = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const db = getDbPool();
    const result = await db.query('SELECT id, name, email, created_at, updated_at FROM users WHERE id = $1', [id]);

    if (result.rows.length === 0) {
      return res.status(404).json({ status: 'error', message: 'User not found' });
    }

    res.json({
      status: 'success',
      data: result.rows[0],
    });
  } catch (err: any) {
    console.error('[UserController] getUserById error:', err);
    res.status(500).json({ status: 'error', message: 'Failed to fetch user' });
  }
};

export const createUser = async (req: Request, res: Response) => {
  try {
    const { name, email } = req.body;
    if (!name || !email) {
      return res.status(400).json({ status: 'error', message: 'Name and email are required fields' });
    }

    const db = getDbPool();
    const result = await db.query(
      'INSERT INTO users (name, email) VALUES ($1, $2) RETURNING id, name, email, created_at, updated_at',
      [name, email]
    );

    res.status(201).json({
      status: 'success',
      data: result.rows[0],
    });
  } catch (err: any) {
    if (err.code === '23505') {
      return res.status(409).json({ status: 'error', message: 'User with this email already exists' });
    }
    console.error('[UserController] createUser error:', err);
    res.status(500).json({ status: 'error', message: 'Failed to create user' });
  }
};

export const deleteUser = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const db = getDbPool();
    const result = await db.query('DELETE FROM users WHERE id = $1 RETURNING id', [id]);

    if (result.rows.length === 0) {
      return res.status(404).json({ status: 'error', message: 'User not found' });
    }

    res.json({
      status: 'success',
      message: `User ${id} deleted successfully`,
    });
  } catch (err: any) {
    console.error('[UserController] deleteUser error:', err);
    res.status(500).json({ status: 'error', message: 'Failed to delete user' });
  }
};
